File: minesweeper/game.py
"""Core minesweeper rules over an arbitrary cell graph.

The board is described by an adjacency mapping ``cell -> neighbors``;
cells can be any hashable ids. Board geometry (squares, triangles,
hexagons, ...) lives in :mod:`minesweeper.boards`.
"""

from __future__ import annotations

import random
from enum import Enum
from typing import Hashable, Iterable, Mapping

Cell = Hashable


class CellState(Enum):
    HIDDEN = "hidden"
    REVEALED = "revealed"
    FLAGGED = "flagged"


class GameState(Enum):
    PLAYING = "playing"
    WON = "won"
    LOST = "lost"


class Game:
    """A single minesweeper game on an arbitrary board.

    Mines are placed lazily on the first reveal so the first click is
    always safe. Deterministic behaviour for tests is available either by
    passing an explicit ``mine_positions`` set or a seeded ``rng``.
    """

    def __init__(
        self,
        adjacency: Mapping[Cell, Iterable[Cell]],
        mine_count: int | None = None,
        *,
        mine_positions: set[Cell] | None = None,
        rng: random.Random | None = None,
    ) -> None:
        self._adjacency = {
            cell: tuple(neighbors) for cell, neighbors in adjacency.items()
        }
        if not self._adjacency:
            raise ValueError("board has no cells")
        for cell, neighbors in self._adjacency.items():
            for neighbor in neighbors:
                if neighbor not in self._adjacency:
                    raise ValueError(
                        f"neighbor {neighbor!r} of {cell!r} is not a board cell"
                    )
        if mine_positions is not None:
            unknown = set(mine_positions) - self._adjacency.keys()
            if unknown:
                raise ValueError(f"mine positions not on the board: {unknown!r}")
            mine_count = len(mine_positions)
        if mine_count is None or mine_count < 1:
            raise ValueError("need at least one mine")
        if mine_count >= len(self._adjacency):
            raise ValueError("mine count must leave at least one safe cell")

        self.mine_count = mine_count
        self.state = GameState.PLAYING
        self._rng = rng if rng is not None else random.Random()
        self._mines: set[Cell] = set(mine_positions or ())
        self._mines_placed = mine_positions is not None
        self._cell_states = {cell: CellState.HIDDEN for cell in self._adjacency}
        self._revealed_count = 0

    # -- queries ----------------------------------------------------------

    @property
    def cells(self) -> tuple[Cell, ...]:
        return tuple(self._adjacency)

    def neighbors(self, cell: Cell) -> tuple[Cell, ...]:
        return self._adjacency[cell]

    def cell_state(self, cell: Cell) -> CellState:
        return self._cell_states[cell]

    def is_mine(self, cell: Cell) -> bool:
        return cell in self._mines

    def adjacent_mines(self, cell: Cell) -> int:
        return sum(1 for n in self._adjacency[cell] if n in self._mines)

    @property
    def flags_remaining(self) -> int:
        flagged = sum(
            1 for state in self._cell_states.values() if state is CellState.FLAGGED
        )
        return self.mine_count - flagged

    # -- moves ------------------------------------------------------------

    def reveal(self, cell: Cell) -> None:
        """Reveal a cell. Ends the game if it is a mine; flood-fills from
        cells with no adjacent mines. No-op on flagged/revealed/unknown
        cells or after the game has ended."""
        if self.state is not GameState.PLAYING or cell not in self._adjacency:
            return
        if self._cell_states[cell] is not CellState.HIDDEN:
            return

        if not self._mines_placed:
            self._place_mines(safe=cell)

        if cell in self._mines:
            self._cell_states[cell] = CellState.REVEALED
            self.state = GameState.LOST
            return

        self._flood_reveal(cell)
        if self._revealed_count == len(self._adjacency) - self.mine_count:
            self.state = GameState.WON
            # flag the remaining mines so the flag counter reads zero
            for mine in self._mines:
                if self._cell_states[mine] is CellState.HIDDEN:
                    self._cell_states[mine] = CellState.FLAGGED

    def toggle_flag(self, cell: Cell) -> None:
        if self.state is not GameState.PLAYING or cell not in self._adjacency:
            return
        current = self._cell_states[cell]
        if current is CellState.HIDDEN:
            self._cell_states[cell] = CellState.FLAGGED
        elif current is CellState.FLAGGED:
            self._cell_states[cell] = CellState.HIDDEN

    def chord(self, cell: Cell) -> None:
        """Reveal all unflagged neighbors of a revealed cell whose flag
        count matches its adjacent mine count."""
        if self.state is not GameState.PLAYING or cell not in self._adjacency:
            return
        if self._cell_states[cell] is not CellState.REVEALED:
            return
        neighbors = self._adjacency[cell]
        flagged = sum(
            1 for n in neighbors if self._cell_states[n] is CellState.FLAGGED
        )
        if flagged != self.adjacent_mines(cell):
            return
        for n in neighbors:
            if self._cell_states[n] is CellState.HIDDEN:
                self.reveal(n)
                if self.state is not GameState.PLAYING:
                    return

    # -- internals --------------------------------------------------------

    def _place_mines(self, safe: Cell) -> None:
        candidates = [cell for cell in self._adjacency if cell != safe]
        self._mines = set(self._rng.sample(candidates, self.mine_count))
        self._mines_placed = True

    def _flood_reveal(self, cell: Cell) -> None:
        stack = [cell]
        while stack:
            current = stack.pop()
            if self._cell_states[current] is not CellState.HIDDEN:
                continue
            self._cell_states[current] = CellState.REVEALED
            self._revealed_count += 1
            if self.adjacent_mines(current) == 0:
                stack.extend(self._adjacency[current])

File: web/src/game.ts
// Port of minesweeper/game.py — core minesweeper rules over an arbitrary cell
// graph (adjacency map). Knows nothing about geometry or rendering.
import type { CellId } from "./boards/core";
import { mulberry32, sample, type Rng } from "./rng";

export type CellState = "hidden" | "revealed" | "flagged";
export type GameState = "playing" | "won" | "lost";

export interface GameOptions {
  mineCount?: number;
  minePositions?: Iterable<CellId>;
  rng?: Rng;
}

export class Game {
  readonly mineCount: number;
  state: GameState = "playing";

  private readonly adjacency: Map<CellId, CellId[]>;
  private readonly rng: Rng;
  private mines: Set<CellId>;
  private minesPlaced: boolean;
  private readonly cellStates = new Map<CellId, CellState>();
  private revealedCount = 0;

  constructor(
    adjacency: Map<CellId, Iterable<CellId>>,
    options: GameOptions = {},
  ) {
    this.adjacency = new Map();
    for (const [cell, neighbors] of adjacency) {
      this.adjacency.set(cell, [...neighbors]);
    }
    if (this.adjacency.size === 0) throw new Error("board has no cells");
    for (const [cell, neighbors] of this.adjacency) {
      for (const n of neighbors) {
        if (!this.adjacency.has(n)) {
          throw new Error(`neighbor ${n} of ${cell} is not a board cell`);
        }
      }
    }

    let mineCount = options.mineCount;
    const explicit = options.minePositions
      ? new Set(options.minePositions)
      : null;
    if (explicit) {
      for (const m of explicit) {
        if (!this.adjacency.has(m)) {
          throw new Error(`mine position not on the board: ${m}`);
        }
      }
      mineCount = explicit.size;
    }
    if (mineCount == null || mineCount < 1) {
      throw new Error("need at least one mine");
    }
    if (mineCount >= this.adjacency.size) {
      throw new Error("mine count must leave at least one safe cell");
    }

    this.mineCount = mineCount;
    this.rng = options.rng ?? mulberry32((Math.random() * 2 ** 32) >>> 0);
    this.mines = explicit ?? new Set();
    this.minesPlaced = explicit != null;
    for (const cell of this.adjacency.keys()) this.cellStates.set(cell, "hidden");
  }

  // -- queries ---------------------------------------------------------------

  get cells(): CellId[] {
    return [...this.adjacency.keys()];
  }

  neighbors(cell: CellId): CellId[] {
    return this.adjacency.get(cell) ?? [];
  }

  cellState(cell: CellId): CellState {
    return this.cellStates.get(cell)!;
  }

  isMine(cell: CellId): boolean {
    return this.mines.has(cell);
  }

  adjacentMines(cell: CellId): number {
    let n = 0;
    for (const neighbor of this.adjacency.get(cell) ?? []) {
      if (this.mines.has(neighbor)) n++;
    }
    return n;
  }

  get flagsRemaining(): number {
    let flagged = 0;
    for (const s of this.cellStates.values()) if (s === "flagged") flagged++;
    return this.mineCount - flagged;
  }

  get revealed(): number {
    return this.revealedCount;
  }

  // -- moves -----------------------------------------------------------------

  /** Reveal a cell; changed cells are returned for ranged rendering updates. */
  reveal(cell: CellId): CellId[] {
    if (this.state !== "playing" || !this.adjacency.has(cell)) return [];
    if (this.cellStates.get(cell) !== "hidden") return [];

    if (!this.minesPlaced) this.placeMines(cell);

    if (this.mines.has(cell)) {
      this.cellStates.set(cell, "revealed");
      this.state = "lost";
      return [cell];
    }

    const changed = this.floodReveal(cell);
    if (this.revealedCount === this.adjacency.size - this.mineCount) {
      this.state = "won";
      for (const mine of this.mines) {
        if (this.cellStates.get(mine) === "hidden") {
          this.cellStates.set(mine, "flagged");
          changed.push(mine);
        }
      }
    }
    return changed;
  }

  toggleFlag(cell: CellId): CellId[] {
    if (this.state !== "playing" || !this.adjacency.has(cell)) return [];
    const current = this.cellStates.get(cell);
    if (current === "hidden") {
      this.cellStates.set(cell, "flagged");
      return [cell];
    }
    if (current === "flagged") {
      this.cellStates.set(cell, "hidden");
      return [cell];
    }
    return [];
  }

  /** Reveal all unflagged neighbours of a revealed cell whose flag count
   * matches its adjacent-mine count. */
  chord(cell: CellId): CellId[] {
    if (this.state !== "playing" || !this.adjacency.has(cell)) return [];
    if (this.cellStates.get(cell) !== "revealed") return [];
    const neighbors = this.adjacency.get(cell)!;
    let flagged = 0;
    for (const n of neighbors) if (this.cellStates.get(n) === "flagged") flagged++;
    if (flagged !== this.adjacentMines(cell)) return [];
    const changed: CellId[] = [];
    for (const n of neighbors) {
      if (this.cellStates.get(n) === "hidden") {
        changed.push(...this.reveal(n));
        if (this.state !== "playing") return changed;
      }
    }
    return changed;
  }

  // -- internals -------------------------------------------------------------

  private placeMines(safe: CellId): void {
    const candidates = this.cells.filter((c) => c !== safe);
    this.mines = new Set(sample(candidates, this.mineCount, this.rng));
    this.minesPlaced = true;
  }

  private floodReveal(cell: CellId): CellId[] {
    const changed: CellId[] = [];
    const stack = [cell];
    while (stack.length) {
      const current = stack.pop()!;
      if (this.cellStates.get(current) !== "hidden") continue;
      this.cellStates.set(current, "revealed");
      this.revealedCount++;
      changed.push(current);
      if (this.adjacentMines(current) === 0) {
        stack.push(...(this.adjacency.get(current) ?? []));
      }
    }
    return changed;
  }
}
